src/vcpkg/base/files.cpp

```diff
@@ -1682,6 +1824,206 @@ namespace vcpkg
+                    errno = 0;
+                    auto entry = op.read();
+                    if (!entry)
+                    {
+                        if (errno != 0)
+                        {
+                            ec.assign(errno, std::generic_category());
+                            return;
+                        }
+
+                        // no more entries left
+                        break;
+                    }
+
+                    if (is_dot_or_dot_dot(entry->d_name))
+                    {
+                        continue;
+                    }
+
+                    auto full = base / entry->d_name;
+                    if (VCPKG_GET_D_TYPE(entry) == VCPKG_DT_UNKNOWN)
+                    {
+                        struct stat s;
+                        if (stat(full.c_str(), &s) != 0 && errno != ENOENT)
+                        {
+                            ec.assign(errno, std::generic_category());
+                            result.clear();
+                            return;
+                        }
+
+                        if (selector(VCPKG_DT_UNKNOWN, s.st_mode))
+                        {
+                            // push results before recursion to get outer entries first
+                            result.push_back(full);
+                        }
+
+                        if (S_ISDIR(s.st_mode))
+                        {
+                            get_files_recursive_impl(result, full, ec, selector);
+                            if (ec)
+                            {
+                                return;
+                            }
+                        }
+                    }
+                    else
+                    {
+                        if (selector(VCPKG_GET_D_TYPE(entry), 0))
+                        {
+                            // push results before recursion to get outer entries first
+                            result.push_back(full);
+                        }
+
+                        if (VCPKG_GET_D_TYPE(entry) == VCPKG_DT_DIR)
+                        {
+                            get_files_recursive_impl(result, full, ec, selector);
+                            if (ec)
+                            {
+                                return;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        // Selector is a function taking (unsigned char dtype, const Path&) and returning bool
+        // (This is similar to the recursive version, but the non-recursive version doesn't need to do stat calls
+        // so selector needs to do them if it wants.)
+        template<class Selector>
+        static void get_files_non_recursive_impl(std::vector<Path>& result,
+                                                 const Path& base,
+                                                 std::error_code& ec,
+                                                 Selector selector)
```

Comment:
I'd flatten Selector to a function pointer instead of a template -- we're touching the disk so a "virtual" call is free and more specific